In [2]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Introduction

Above, is the data about online stall in eBay owned by "Privat" that sells mostly used cars. we have 50.000 entries and most of the columns are strings. according to our findings there are several null data within the dataframe

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:

# mapping_dict={ 'dateCrawled':'date_crawled',
#               'name':'name',
#               'seller':'seller',
#               'offerType':'offer_type',
#               'price':'price',
#               'abtest':'abtest',
#               'vehicleType':'vehicle_type',
#               'yearOfRegistration':'registration_year',
#               'gearbox':'gearbox',
#               'powerPS':'power_ps',
#               'model':'model',
#               'odometer':'odometer',
#               'monthOfRegistration':'registration_month',
#               'fuelType':'fuel_type', 
#               'brand':'brand',
#               'notRepairedDamage':'unrepaired_damage',
#               'dateCreated':'ad_created',
#               'nrOfPictures':'nr_of_pictures',
#               'postalCode':'postal_code',
#               'lastSeen':'last_seen'       
    
# }
# autos_columns = autos.columns.map(mapping_dict)
# autos.columns = autos_columns
# autos.head()

In [8]:

import re

def camel_to_snake(name):
    # Use regular expression to convert camelcase to snakecase
    name = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
    return re.sub('([a-z0-9])([A-Z])', r'\1_\2', name).lower()

In [9]:
new_cols = []
for i in autos.columns:
    new_cols.append(camel_to_snake(i))
autos.columns = new_cols

In [10]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  year_of_registration  \
0  privat    Angebot  $5,000  control          bus                  2004   
1  privat    Angebot  $8,500  control    limousine                  1997   
2  privat    Angebot  $8,990     test    limousine                  2009   
3  privat    Angebot  $4,350  control   kleinwagen                  2007   
4  privat    Angebot  $1,350     test        kombi                  2003   

     gearbox  power_ps   model   odometer  month_of_registration fuel_type  \
0    manuell       158  andere  150,000km                      3       lpg   
1  automatik       286     7er  150,000km                      6    benzin   
2    manuell       102    golf   70,000km                      7    benzin   
3  automatik        71  fortwo   70,000km                      6    benzin   
4    manuell         0   focus  150,000km                      7    benzin   

        brand not_repaired_damage         date_created  nr_of_pictures  \
0     peugeot                nein  2016-03-26 00:00:00               0   
1         bmw                nein  2016-04-04 00:00:00               0   
2  volkswagen                nein  2016-03-26 00:00:00               0   
3       smart                nein  2016-03-12 00:00:00               0   
4        ford                nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Chaging Camelcase to Snakecase

As using camelcase for column name is not the best practing. Above, we create a function that able to convert camelcase string as an input and return it as snake_case. we then implement the case to the autos dataframe column name. 

In [11]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 15:49:30  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  year_of_registration  gearbox      power_ps  model  \
count         44905          50000.000000    47320  50000.000000  47242   
unique            8                   NaN        2           NaN    245   
top       limousine                   NaN  manuell           NaN   golf   
freq          12859                   NaN    36993           NaN   4024   
mean            NaN           2005.073280      NaN    116.355920    NaN   
std             NaN            105.712813      NaN    209.216627    NaN   
min             NaN           1000.000000      NaN      0.000000    NaN   
25%             NaN           1999.000000      NaN     70.000000    NaN   
50%             NaN           2003.000000      NaN    105.000000    NaN   
75%             NaN           2008.000000      NaN    150.000000    NaN   
max             NaN           9999.000000      NaN  17700.000000    NaN   

         odometer  month_of_registration fuel_type       brand  \
count       50000           50000.000000     45518       50000   
unique         13                    NaN         7          40   
top     150,000km                    NaN    benzin  volkswagen   
freq        32424                    NaN     30107       10687   
mean          NaN               5.723360       NaN         NaN   
std           NaN               3.711984       NaN         NaN   
min           NaN               0.000000       NaN         NaN   
25%           NaN               3.000000       NaN         NaN   
50%           NaN               6.000000       NaN         NaN   
75%           NaN               9.000000       NaN         NaN   
max           NaN              12.000000       NaN         NaN   

       not_repaired_damage         date_created  nr_of_pictures   postal_code  \
count                40171                50000         50000.0  50000.000000   
unique                   2                   76             NaN           NaN   
top                   nein  2016-04-03 00:00:00             NaN           NaN   
freq                 35232                 1946             NaN           NaN   
mean                   NaN                  NaN             0.0  50813.627300   
std                    NaN                  NaN             0.0  25779.747957   
min                    NaN                  NaN             0.0   1067.000000   
25%                    NaN                  NaN             0.0  30451.000000   
50%                    NaN                  NaN             0.0  49577.000000   
75%                    NaN                  NaN             0.0  71540.000000   
max                    NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Based on the provided data, here are some observations:

**Columns that have mostly one value and are candidates to be dropped:**

`"seller"`: contains mostly one value "privat" (49999 out of 50000).

`"offer_type"`: contains mostly one value "Angebot" (49999 out of 50000).

`"ab_test"`: contains mostly one value "test" (25756 out of 50000).

`"gear_box"`: contains mostly one value "manuell" (36993 out of 50000).

`"unrepaired_damage"`: contains mostly one value "nein" (35232 out of 50000).

**Columns that need more investigation:**

`"name"`: contains many unique values (38754 out of 50000), so it's difficult to draw any conclusions without more information on what this column represents.

`"vehicle_type"`: missing values (5000 out of 50000) and contains multiple categories that need further investigation.

`"model"`: contains missing values (2758 out of 50000) and many different categories that need further investigation.
`"odometer"`: contains numeric data stored as text (e.g. "150,000km"). It needs to be cleaned and converted to numeric data.

`"fuel_type"`: missing values (4482 out of 50000) and multiple categories that need further investigation.

`"ad_created"`: contains many unique values (76 out of 50000), so it's difficult to draw any conclusions without more information on what this column represents.

`"nr_of_pictures"`: contains mostly one value "0" (50000 out of 50000), so it is a candidate to be dropped.

`"postal_code"`: contains missing values (0 out of 50000), but further investigation may reveal interesting information.

After further investigation, we can see that the `"registration_year"` column contains unrealistic values (e.g. 1000 and 9999), indicating errors in the data. These rows should be removed or corrected. Finally, the `"vehicle_type"`, `"model"`, `"registration_month"`, `"fuel_type"`, and `"ad_created"` columns require further investigation.


In [12]:
autos['odometer']

0        150,000km
1        150,000km
2         70,000km
3         70,000km
4        150,000km
           ...    
49995    100,000km
49996    150,000km
49997      5,000km
49998     40,000km
49999    150,000km
Name: odometer, Length: 50000, dtype: object

In [13]:
autos['price']

0         $5,000
1         $8,500
2         $8,990
3         $4,350
4         $1,350
          ...   
49995    $24,900
49996     $1,980
49997    $13,200
49998    $22,900
49999     $1,250
Name: price, Length: 50000, dtype: object

In [14]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   date_crawled           50000 non-null  object
 1   name                   50000 non-null  object
 2   seller                 50000 non-null  object
 3   offer_type             50000 non-null  object
 4   price                  50000 non-null  object
 5   abtest                 50000 non-null  object
 6   vehicle_type           44905 non-null  object
 7   year_of_registration   50000 non-null  int64 
 8   gearbox                47320 non-null  object
 9   power_ps               50000 non-null  int64 
 10  model                  47242 non-null  object
 11  odometer               50000 non-null  object
 12  month_of_registration  50000 non-null  int64 
 13  fuel_type              45518 non-null  object
 14  brand                  50000 non-null  object
 15  not_repaired_damage

In [15]:

autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(float)
autos['odometer'] = autos['odometer'].str.replace('km','').str.replace(',','').astype(float)
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)
autos.rename({'price': 'price_usd'}, axis=1, inplace=True)                                                                              

In [16]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   date_crawled           50000 non-null  object 
 1   name                   50000 non-null  object 
 2   seller                 50000 non-null  object 
 3   offer_type             50000 non-null  object 
 4   price_usd              50000 non-null  float64
 5   abtest                 50000 non-null  object 
 6   vehicle_type           44905 non-null  object 
 7   year_of_registration   50000 non-null  int64  
 8   gearbox                47320 non-null  object 
 9   power_ps               50000 non-null  int64  
 10  model                  47242 non-null  object 
 11  odometer_km            50000 non-null  float64
 12  month_of_registration  50000 non-null  int64  
 13  fuel_type              45518 non-null  object 
 14  brand                  50000 non-null  object 
 15  no

## Exploring the Odometer and Price Columns


### price

In [17]:
autos['price_usd'].unique().shape

(2357,)

In [18]:
autos['price_usd'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

In [19]:
autos['price_usd'].value_counts()

0.0        1421
500.0       781
1500.0      734
2500.0      643
1200.0      639
           ... 
6202.0        1
18310.0       1
898.0         1
11240.0       1
789.0         1
Name: price_usd, Length: 2357, dtype: int64

In [20]:
autos["price_usd"].value_counts().sort_index(ascending=True).head(20)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
Name: price_usd, dtype: int64

In [21]:
autos['price_usd'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

In [22]:
autos["price_usd"].value_counts().sort_index(ascending=False).head(20)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price_usd, dtype: int64

In [23]:
autos = autos[(autos['price_usd']>=500) & (autos['price_usd']<=1000000)]

In [24]:
autos["price_usd"].value_counts().sort_index(ascending=False).head(20)

999999.0    2
999990.0    1
350000.0    1
345000.0    1
299000.0    1
295000.0    1
265000.0    1
259000.0    1
250000.0    1
220000.0    1
198000.0    1
197000.0    1
194000.0    1
190000.0    1
180000.0    1
175000.0    1
169999.0    1
169000.0    1
163991.0    1
163500.0    1
Name: price_usd, dtype: int64

we found there are many cars with pricing that is uncommon. we assume the minimum price for a car is 500 USD and max car that is sold in a ecommerce platform will be 1.000.000 USD max.

### odometer

In [25]:
autos['odometer_km'].unique().shape

(13,)

In [26]:
autos['odometer_km'].unique()

array([150000.,  70000.,  50000.,  80000.,  10000.,  30000., 125000.,
        90000.,  20000.,  60000.,   5000., 100000.,  40000.])

we believe that we don't find any outliers on odometer columns

In [27]:
autos.shape

(45100, 20)

## Analyzing Date

In [28]:
autos.describe()

price_usd  year_of_registration      power_ps    odometer_km  \
count   45100.000000          45100.000000  45100.000000   45100.000000   
mean     6386.757738           2005.063282    121.557539  125294.124169   
std     12306.631453             89.649158    205.721969   39621.768672   
min       500.000000           1000.000000      0.000000    5000.000000   
25%      1500.000000           2000.000000     75.000000  100000.000000   
50%      3500.000000           2004.000000    110.000000  150000.000000   
75%      7900.000000           2008.000000    150.000000  150000.000000   
max    999999.000000           9999.000000  17700.000000  150000.000000   

       month_of_registration  nr_of_pictures   postal_code  
count           45100.000000         45100.0  45100.000000  
mean                5.907073             0.0  51291.163991  
std                 3.632560             0.0  25701.276222  
min                 0.000000             0.0   1067.000000  
25%                 3.000000             0.0  30962.500000  
50%                 6.000000             0.0  50226.000000  
75%                 9.000000             0.0  72160.000000  
max                12.000000             0.0  99998.000000

In [29]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45100 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   date_crawled           45100 non-null  object 
 1   name                   45100 non-null  object 
 2   seller                 45100 non-null  object 
 3   offer_type             45100 non-null  object 
 4   price_usd              45100 non-null  float64
 5   abtest                 45100 non-null  object 
 6   vehicle_type           41425 non-null  object 
 7   year_of_registration   45100 non-null  int64  
 8   gearbox                43286 non-null  object 
 9   power_ps               45100 non-null  int64  
 10  model                  43035 non-null  object 
 11  odometer_km            45100 non-null  float64
 12  month_of_registration  45100 non-null  int64  
 13  fuel_type              41901 non-null  object 
 14  brand                  45100 non-null  object 
 15  no

In [30]:
(autos['date_crawled']
        .str[:10]   #Selects the first 10 characters from each column
        .value_counts(normalize=True, dropna=False)  #Returns unique values with their proportions and also includes missing values
        .sort_index() #Returns data in ascending order
        )

2016-03-05    0.025565
2016-03-06    0.014124
2016-03-07    0.036186
2016-03-08    0.033171
2016-03-09    0.032905
2016-03-10    0.032705
2016-03-11    0.033016
2016-03-12    0.037317
2016-03-13    0.015521
2016-03-14    0.036297
2016-03-15    0.034013
2016-03-16    0.029357
2016-03-17    0.031175
2016-03-18    0.012882
2016-03-19    0.034745
2016-03-20    0.038071
2016-03-21    0.037761
2016-03-22    0.033016
2016-03-23    0.032395
2016-03-24    0.028980
2016-03-25    0.031086
2016-03-26    0.032639
2016-03-27    0.031175
2016-03-28    0.034834
2016-03-29    0.033282
2016-03-30    0.033326
2016-03-31    0.031663
2016-04-01    0.033902
2016-04-02    0.035765
2016-04-03    0.038825
2016-04-04    0.036608
2016-04-05    0.013171
2016-04-06    0.003171
2016-04-07    0.001353
Name: date_crawled, dtype: float64

In [31]:
(autos['date_created']
        .str[:10]   #Selects the first 10 characters from each column
        .value_counts(normalize=True, dropna=False)  #Returns unique values with their proportions and also includes missing values
        .sort_index() #Returns data in ascending order
        )

2015-06-11    0.000022
2015-08-10    0.000022
2015-09-09    0.000022
2015-11-10    0.000022
2015-12-05    0.000022
                ...   
2016-04-03    0.039047
2016-04-04    0.036984
2016-04-05    0.011907
2016-04-06    0.003259
2016-04-07    0.001197
Name: date_created, Length: 76, dtype: float64

In [32]:
(autos['last_seen']
        .str[:10]   #Selects the first 10 characters from each column
        .value_counts(normalize=True, dropna=False)  #Returns unique values with their proportions and also includes missing values
        .sort_index() #Returns data in ascending order
        )

2016-03-05    0.001086
2016-03-06    0.004169
2016-03-07    0.005211
2016-03-08    0.007007
2016-03-09    0.009468
2016-03-10    0.010288
2016-03-11    0.012040
2016-03-12    0.023902
2016-03-13    0.008869
2016-03-14    0.012284
2016-03-15    0.015676
2016-03-16    0.016164
2016-03-17    0.027672
2016-03-18    0.007406
2016-03-19    0.015410
2016-03-20    0.020421
2016-03-21    0.020665
2016-03-22    0.021242
2016-03-23    0.018404
2016-03-24    0.019534
2016-03-25    0.018581
2016-03-26    0.016475
2016-03-27    0.015455
2016-03-28    0.020554
2016-03-29    0.021375
2016-03-30    0.024146
2016-03-31    0.023437
2016-04-01    0.022860
2016-04-02    0.024878
2016-04-03    0.024945
2016-04-04    0.024302
2016-04-05    0.126608
2016-04-06    0.225322
2016-04-07    0.134146
Name: last_seen, dtype: float64

In [33]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45100 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   date_crawled           45100 non-null  object 
 1   name                   45100 non-null  object 
 2   seller                 45100 non-null  object 
 3   offer_type             45100 non-null  object 
 4   price_usd              45100 non-null  float64
 5   abtest                 45100 non-null  object 
 6   vehicle_type           41425 non-null  object 
 7   year_of_registration   45100 non-null  int64  
 8   gearbox                43286 non-null  object 
 9   power_ps               45100 non-null  int64  
 10  model                  43035 non-null  object 
 11  odometer_km            45100 non-null  float64
 12  month_of_registration  45100 non-null  int64  
 13  fuel_type              41901 non-null  object 
 14  brand                  45100 non-null  object 
 15  no

we found no null values on all the date_crawled, date_created, last_seen.

In [36]:
autos['year_of_registration'].describe()

count    45100.000000
mean      2005.063282
std         89.649158
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: year_of_registration, dtype: float64

it is found that most of the cars that being sold were, from year 2000s. but There are some outliers data that need to be taken out

In [38]:
autos['year_of_registration'].value_counts().sort_index(ascending=False).head(20)

9999       3
9000       1
8888       1
6200       1
5911       1
5000       2
4800       1
4500       1
4100       1
2800       1
2019       1
2018     462
2017    1296
2016     950
2015     364
2014     652
2013     797
2012    1307
2011    1617
2010    1586
Name: year_of_registration, dtype: int64

In [40]:
autos['date_created'].value_counts().sort_index(ascending=False).head(20)

2016-04-07 00:00:00      54
2016-04-06 00:00:00     147
2016-04-05 00:00:00     537
2016-04-04 00:00:00    1668
2016-04-03 00:00:00    1761
2016-04-02 00:00:00    1599
2016-04-01 00:00:00    1526
2016-03-31 00:00:00    1431
2016-03-30 00:00:00    1497
2016-03-29 00:00:00    1500
2016-03-28 00:00:00    1573
2016-03-27 00:00:00    1404
2016-03-26 00:00:00    1472
2016-03-25 00:00:00    1409
2016-03-24 00:00:00    1306
2016-03-23 00:00:00    1453
2016-03-22 00:00:00    1478
2016-03-21 00:00:00    1713
2016-03-20 00:00:00    1723
2016-03-19 00:00:00    1516
Name: date_created, dtype: int64

looking that the ad was created on 2016, it's not possible to have a car that is registered above 2016.

In [41]:
autos['year_of_registration'].value_counts().sort_index(ascending=True).head(20)

1000    1
1001    1
1910    2
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
1950    1
1951    2
1952    1
1953    1
1954    2
1955    2
1956    4
Name: year_of_registration, dtype: int64

In [42]:
# we will set minimum value 1910 and max value 2018
autos = autos[(autos['year_of_registration']>=1910)&(autos['year_of_registration']<=2018)]

In [43]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45084 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   date_crawled           45084 non-null  object 
 1   name                   45084 non-null  object 
 2   seller                 45084 non-null  object 
 3   offer_type             45084 non-null  object 
 4   price_usd              45084 non-null  float64
 5   abtest                 45084 non-null  object 
 6   vehicle_type           41425 non-null  object 
 7   year_of_registration   45084 non-null  int64  
 8   gearbox                43277 non-null  object 
 9   power_ps               45084 non-null  int64  
 10  model                  43023 non-null  object 
 11  odometer_km            45084 non-null  float64
 12  month_of_registration  45084 non-null  int64  
 13  fuel_type              41894 non-null  object 
 14  brand                  45084 non-null  object 
 15  no

## Explore Price by Brand

In [44]:
autos['brand'].unique

<bound method Series.unique of 0           peugeot
1               bmw
2        volkswagen
3             smart
4              ford
            ...    
49995          audi
49996          opel
49997          fiat
49998          audi
49999          opel
Name: brand, Length: 45084, dtype: object>

In [46]:
#Selecting the top 20 brands
top_brands = autos["brand"].value_counts(normalize=True).head(20).index
top_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'smart', 'mazda', 'nissan',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini'],
      dtype='object')

In [48]:
brand_mean_prices = {}    #Empty dictionary

#Looping through the top 20 brands
for brand in top_brands:
    selected_rows = autos[autos["brand"] == brand]  #This line selects rows from the autos DataFrame where the "brand" column matches the current brand being processed in the loop. The result is stored in the selected_rows variable.
    mean_price = selected_rows["price_usd"].mean()  #This line calculates the mean price of the selected rows. The resulting mean price is stored in the mean_price variable.
    brand_mean_prices[brand] = mean_price   #This line adds an entry to the brand_mean_prices dictionary. The key for the new entry is the current brand being processed in the loop, and the value is the mean_price calculated in the previous line. The result is a dictionary that maps each brand to its mean price.


brand_mean_prices

{'volkswagen': 5908.072674418605,
 'bmw': 8514.81103286385,
 'opel': 3345.5830967460142,
 'mercedes_benz': 8660.896853146853,
 'audi': 9484.499876390606,
 'ford': 4607.311923471131,
 'renault': 2760.8114143920598,
 'peugeot': 3329.3254211332314,
 'fiat': 3211.992606284658,
 'seat': 4709.527446300716,
 'skoda': 6539.7107001321,
 'smart': 3552.050946142649,
 'mazda': 4391.736377025037,
 'nissan': 5077.141592920354,
 'citroen': 3988.830218068536,
 'toyota': 5244.499165275459,
 'hyundai': 5695.464757709251,
 'sonstige_autos': 13272.767441860466,
 'volvo': 5110.575539568345,
 'mini': 10616.96626506024}

Analyzing the results, we find that there is a significant variation in the mean prices across the top 20 brands. The top three brands with the highest mean prices are Porsche, Land Rover, and Audi, while the lowest mean prices are for Renault, Peugeot, and Opel. This suggests that brand plays an important role in determining the price of a car in the secondary market.

## Storing Aggregate Data in a DataFrame

In [49]:
bmp_series = pd.Series(brand_mean_prices)

In [51]:
bmp_series.head(6)

volkswagen       5908.072674
bmw              8514.811033
opel             3345.583097
mercedes_benz    8660.896853
audi             9484.499876
ford             4607.311923
dtype: float64

In [52]:
brand_mean_mileage = {}   #empty dictionary

#aggregating data
for brand in top_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

#Converting both dictionaries to series object
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)  #Creating a series and storing in a variable
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [53]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])   #Creating a dataframe from the series objct
brand_info

mean_mileage
volvo                 138992
bmw                   132970
mercedes_benz         131074
audi                  129283
volkswagen            128513
opel                  128199
renault               126493
peugeot               126271
mazda                 124219
ford                  123647
seat                  120763
citroen               118956
nissan                117352
toyota                115843
fiat                  115032
skoda                 111010
hyundai               105253
smart                 100705
sonstige_autos         91337
mini                   89469

In [54]:
brand_info["mean_price"] = bmp_series  #Assigning the other series as a new column in the newly created dataframe
brand_info

mean_mileage    mean_price
volvo                 138992   5110.575540
bmw                   132970   8514.811033
mercedes_benz         131074   8660.896853
audi                  129283   9484.499876
volkswagen            128513   5908.072674
opel                  128199   3345.583097
renault               126493   2760.811414
peugeot               126271   3329.325421
mazda                 124219   4391.736377
ford                  123647   4607.311923
seat                  120763   4709.527446
citroen               118956   3988.830218
nissan                117352   5077.141593
toyota                115843   5244.499165
fiat                  115032   3211.992606
skoda                 111010   6539.710700
hyundai               105253   5695.464758
smart                 100705   3552.050946
sonstige_autos         91337  13272.767442
mini                   89469  10616.966265

## Next Steps Analysis

In [55]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price_usd', 'abtest',
       'vehicle_type', 'year_of_registration', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'month_of_registration', 'fuel_type', 'brand',
       'not_repaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [56]:
# Identify categorical data that uses German words, translate them and map the values to their English counterparts
# For example, translating 'privat' to 'private' and 'gewerblich' to 'commercial'

autos['seller'] = autos['seller'].map({'privat': 'private', 'gewerblich': 'commercial'})
autos['offer_type'] = autos['offer_type'].map({'Angebot': 'offer', 'Gesuch': 'request'})
autos['gearbox'] = autos['gearbox'].map({'manuell': 'manual', 'automatik': 'automatic'})
autos['not_repaired_damage'] = autos['not_repaired_damage'].map({'ja': 'yes', 'nein': 'no'})

In [57]:
# Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.

autos['date_crawled'] = pd.to_datetime(autos['date_crawled'], format='%Y-%m-%d').dt.strftime('%Y%m%d').astype(int)
autos['ad_created'] = pd.to_datetime(autos['date_created'], format='%Y-%m-%d').dt.strftime('%Y%m%d').astype(int)
autos['last_seen'] = pd.to_datetime(autos['last_seen'], format='%Y-%m-%d').dt.strftime('%Y%m%d').astype(int)

In [58]:
brand_model = autos.groupby(['model']).size().reset_index(name='count')
brand_model_sorted = brand_model.sort_values('count', ascending=False)
brand_model_sorted.head(10)

model  count
116      golf   3621
40     andere   3211
11        3er   2583
170      polo   1416
84      corsa   1341
167    passat   1341
43      astra   1284
29         a4   1223
60   c_klasse   1147
15        5er   1132

In [59]:
autos['odometer_km'].value_counts()

150000.0    28699
125000.0     4839
100000.0     2031
90000.0      1676
80000.0      1385
70000.0      1189
60000.0      1131
50000.0       996
40000.0       808
30000.0       765
20000.0       727
5000.0        608
10000.0       230
Name: odometer_km, dtype: int64

In [61]:
# Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.
autos['mileage_group'] = pd.cut(autos['odometer_km'], bins=[0, 50000, 100000, 150000, 200000, np.inf], labels=['<50k', '50k-100k', '100k-150k', '150k-200k', '>200k'])
mileage_prices = autos.groupby('mileage_group')['price_usd'].mean().reset_index(name='avg_price')
mileage_prices

mileage_group     avg_price
0          <50k  15771.279874
1      50k-100k   9765.319077
2     100k-150k   4482.139961
3     150k-200k           NaN
4         >200k           NaN

In [65]:
# How much cheaper are cars with damage than their non-damaged counterparts?

damaged_prices = autos[autos['not_repaired_damage'] == 'yes']['price_usd'].mean()
undamaged_prices = autos[autos['not_repaired_damage'] == 'no']['price_usd'].mean()
price_diff = (undamaged_prices - damaged_prices) / undamaged_prices * 100

print('Undamaged cars are, on average, {:.2f}% more expensive than damaged cars'.format(price_diff))

Undamaged cars are, on average, 58.42% more expensive than damaged cars
